Fix offline hotplug test failing

awels opened the pull request:
There is a race condition when removing multiple volumes at once.
It is possible the first volume gets removed by the virt-controller
while we are calculating the patch for the second volume and now
the patch will fail. This retries by reloading the VM and recalculating
the patch.

Signed-off-by: Alexander Wels <awels@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4974 

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

rmohr commented:
I think that you should add a json-patch `test` operation before, like you do in the controller already.
Sadly json patch return codes are bad, but in the error message you should see something like `test operation does not apply` which will help you identifying that the array content changed and that you should retry.

awels commented:
So the code is already doing the test, I think that is what is failing due to the race condition. I refactored the code that sends the patch a little to have fewer duplicates, and I return the source error if IsInvalid(err) is true. Otherwise its a 500 like before. I also updated the test to retry on failure, now remove matches add

awels commented:
This should get the test out of Quarantine, do I remove the flag in this PR or a separate one?

brybacki commented:
> This should get the test out of Quarantine, do I remove the flag in this PR or a separate one?

The flag will be removed when we confirm test is no longer flaky, a separate PR, that needs to wait for two weeks of stabilization.

awels commented:
@rmohr Can you review?

rmohr commented:
/lgtm
/approve

rmohr commented on tests/storage/hotplug.go:
```diff
@@ -142,28 +142,33 @@ var _ = SIGDescribe("Hotplug", func() {
-		err = removeVolumeCommand()
+		Eventually(func() error {
+			return removeVolumeCommand()
+		}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
```
> That shounds a little bit extreme. Le'ts go with 3 retries or so.

/lgtm cancel

let's adjust this a little

awels replied:
So I did the same with addvolume, want me to adjust that one as well?

awels replied:
Okay adjusted them all to 3 retries.

rmohr commented:
/lgtm
/retest
